Final - New Events Page

nabramow opened the pull request:
This is the final PR for the new events feature. The individual pieces have all already been reviewed in smaller PRs, so here we're looking for any final design or usability issues before it goes to prod.

Closes #4823 

**Web frontend checklist**
- [ x ] Formatted my code with `yarn format`
- [ x ] There are no warnings from `yarn lint --fix`
- [ ] There are no console warnings when running the app
- [ ] Added any new components to storybook
- [ x ] Added tests where relevant
- [ x ] All tests pass
- [ x ] Clicked around my changes running locally and it works
- [ x ] Checked Desktop, Mobile and Tablet screen sizes

nabramow committed:
Make event type toggle and header

nabramow committed:
Fix button padding

nabramow committed:
Create event card skeleton

nabramow committed:
Finish design

nabramow committed:
Add color t Pill and remove endTime

nabramow committed:
Fetch real events

nabramow committed:
Add listMyEvents hook and connect create event button

nabramow committed:
Add discover events section

nabramow committed:
Fix section headings in past view

nabramow committed:
Add objectFit to EventItem for image

nabramow committed:
Design changes for long title

nabramow committed:
Merge feature branch

nabramow committed:
Merge pull request #4863 from Couchers-org/skeleton

New Events Page - DESIGN ONLY

nabramow committed:
Merge remote-tracking branch 'origin/web/feature/new-events-page' into web/feature/new-events-page

nabramow committed:
Change event type toggle to tabs

nabramow committed:
Merge pull request #4882 from Couchers-org/web/feature/event-tabs

Change event type toggle to tabs - DESIGN ONLY

nabramow committed:
Merge in changes from feature branch

nabramow committed:
Add pagination and show cancelled functionality

nabramow committed:
Merge branch 'develop' into web/feature/new-events-page

nabramow committed:
Merge branch 'web/feature/new-events-page' into eventsLogic

nabramow committed:
Add some search filters

nabramow committed:
WIP

nabramow committed:
Merge branch 'develop' into web/feature/new-events-page

nabramow committed:
Merge branch 'web/feature/new-events-page' into eventsLogic

nabramow committed:
Add search by location

nabramow committed:
Merge branch 'develop' into web/feature/new-events-page

nabramow committed:
Merge branch 'web/feature/new-events-page' into eventsLogic

nabramow committed:
Add canceled to discover and clean up

nabramow committed:
Pass showCancelled to LisAllEventsReq

nabramow committed:
Adjust for mobile and center loading spinners

nabramow committed:
Add near me filter

nabramow committed:
Fix mouseover pointer

nabramow committed:
Merge branch 'develop' into web/feature/new-events-page

nabramow committed:
Merge branch 'web/feature/new-events-page' into eventsLogic

nabramow committed:
Review changes except pagination

nabramow committed:
My events pills should not effect discover pills

nabramow committed:
Merge branch 'develop' into web/feature/new-events-page

nabramow committed:
Merge branch 'web/feature/new-events-page' into eventsLogic

nabramow committed:
Add pagination and reorganize component composition for it

nabramow committed:
Fix some responsiveness issues with EventItem

nabramow committed:
Merge pull request #4912 from Couchers-org/eventsLogic

New Events Page - Functionality

nabramow committed:
Add tests for MyEventsList

nabramow committed:
Write tests for DiscoverEventsList

nabramow committed:
Merge branch 'develop' into web/feature/new-events-page

nabramow committed:
Merge branch 'web/feature/new-events-page' into tests-etc

nabramow committed:
Add new events components to existing events folder and rename some

nabramow committed:
More cleanup

nabramow committed:
Remove dependency I added

nabramow committed:
Merge pull request #5003 from Couchers-org/tests-etc

New Events Page - tests and move folders

jesseallhands commented:
Should we change the "communities" and "online" pills to "in-person" and "online"? That's the primary differentiator, right? I don't think we have any in-person events that wouldn't be associated with a community.

![Screenshot_20241019_103726_Firefox](https://github.com/user-attachments/assets/5a3c34fa-a21c-4866-ae94-f5cc5ad34e5e)

bakeiro commented:
> Should we change the "communities" and "online" pills to "in-person" and "online"? That's the primary differentiator, right? I don't think we have any in-person events that wouldn't be associated with a community.

I think there is a difference, because the community pill shows the events only of the communities that I belong, not just any community

jesseallhands commented:
>I think there is a difference, because the community pill shows the events only of the communities that I belong, not just any community

In that case the pill should be "my community" and not just "community".

bakeiro commented:
> > I think there is a difference, because the community pill shows the events only of the communities that I belong, not just any community
> 
> In that case the pill should be "my community" and not just "community".

gotcha... guess that `my communities` it's even better maybe? :D

bakeiro commented:
not the UI guy here but
![image](https://github.com/user-attachments/assets/b2484930-cd33-4beb-bbaa-fdc4f6960c43)

I feels this text a bit small... right? (page wasn't zoomed out), if you feel the same, maybe we can increase it :)

bakeiro commented:
about the code, was something new added to review? (this is just the merged of the reviewed PRs right?)

nabramow commented:
> about the code, was something new added to review? (this is just the merged of the reviewed PRs right?)

No, no new functionality! This is just where all those PRs merged into.

bakeiro commented:
[suggestion here, no needed to approve the PR or something like that]

I was thinking about the events component in the community page... maybe makes sense to standarize these components? (make them more similar, re-using the same component...)

nabramow committed:
Merge branch 'develop' into web/feature/new-events-page

nabramow commented:
> [suggestion here, no needed to approve the PR or something like that]
> 
> I was thinking about the events component in the community page... maybe makes sense to standarize these components? (make them more similar, re-using the same component...)

I agree! Let's do it in a separate ticket? I'd want more feedback about how the pagination should work for that and any filters, and I think this PR is big enough.

nabramow committed:
Merge remote-tracking branch 'origin/web/feature/new-events-page' into web/feature/new-events-page

nabramow committed:
Review changes

nabramow committed:
Fix test

nabramow commented:
Okay I think I addressed everything now!

- Changed 'Communities' -> 'My Communities' @jesseallhands 
- Made the font size for the event info a little bit bigger @bakeiro 
- Put the pills rendering logic in the body of the component instead of a render function @darrenvong 

Is there anything else?

bakeiro commented:
I'm just wondering...

how the event location search works? its by area? (2 coordinates and then search ~50km around it?) if that's the case what happens when user types a country/state/region, would be only the center of that region? (circle area instead of a polygon)

if that's the case we should allow only filter by city... (or smaller 'nodes') I can create another ticket for this if it's needed

let me research about this and create a proper ticket... I already faced similar issues while working on the map

bakeiro commented:
> I'm just wondering...
> 
> how the event location search works? its by area? (2 coordinates and then search ~50km around it?) if that's the case what happens when user types a country/state/region, would be only the center of that region? (circle area instead of a polygon)
> 
> if that's the case we should allow only filter by city... (or smaller 'nodes') I can create another ticket for this if it's needed
> 
> let me research about this and create a proper ticket... I already faced similar issues while working on the map

BTW, this doens't block the PR, since it's already too big

darrenvong commented:
I'm not quite sure how it's related... but on the prod couchers site, I noticed the "remember me" switch label doesn't render on a different line like in this branch:
![Screenshot 2024-10-22 at 01 05 09](https://github.com/user-attachments/assets/9447ff9a-93bd-4e0d-8847-b31f5f0c04c1)

darrenvong commented:
One more quirk I noticed - it's unexpected to me that when you got a location filter on (by typing something in the input), then adding and removing "my communities" and "online" filters would reset it. See recording:

https://github.com/user-attachments/assets/a2f1c96c-0408-4a4e-bf9a-19edf4cae171

In the scenario shown, when I click back to filter on events based on "my communities", I would expect it to only show the event in Sydney since that's what I got in the location input still filled in.

Interestingly, pressing only the searching glass button is enough to put the filter back on without needing to trigger another nomatim search -> select it from the location list before it doing anything, which suggests me that the filter might be still stored in state somewhere but not correctly sent as filters to the backend with the way I used it?

https://github.com/user-attachments/assets/0f3514c3-8c27-47d6-aec5-76810ac2f767

nabramow committed:
Merge branch 'develop' into web/feature/new-events-page

nabramow commented:
> I'm just wondering...
> 
> how the event location search works? its by area? (2 coordinates and then search ~50km around it?) if that's the case what happens when user types a country/state/region, would be only the center of that region? (circle area instead of a polygon)
> 
> if that's the case we should allow only filter by city... (or smaller 'nodes') I can create another ticket for this if it's needed
> 
> let me research about this and create a proper ticket... I already faced similar issues while working on the map

So if you look at `EventSearchReq` or you look at the backend function `def EventSearch` there's a few options for how search can work.

@bakeiro You can "search in rectangle", that's what I chose here. With than you give it a `RectArea` which is a latitude min, latitude max, longitude min and longitude max to search in. So you can calculate that by taking the exact location of the search coordinates and adding some "buffer" around them.

You can also "search in area", with that one you give a latitude, longitude and then a radius.

I don't think it's bad to also allow filter by country or state? What if you live in a tiny country like Monaco or something, haha, or a small US state. Maybe you're willing to drive a bit if you're not in a city? What's the downside or making it flexible?

nabramow committed:
Undo CustomColorSwitch changes since not using

nabramow commented:
> I'm not quite sure how it's related... but on the prod couchers site, I noticed the "remember me" switch label doesn't render on a different line like in this branch: ![Screenshot 2024-10-22 at 01 05 09](https://private-user-images.githubusercontent.com/13669362/378629973-9447ff9a-93bd-4e0d-8847-b31f5f0c04c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2NjAwMTksIm5iZiI6MTcyOTY1OTcxOSwicGF0aCI6Ii8xMzY2OTM2Mi8zNzg2Mjk5NzMtOTQ0N2ZmOWEtOTNiZC00ZTBkLTg4NDctYjMxZjVmMGMwNGMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIzVDA1MDE1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5OWU5NWM3NGM3ZGNlOTFlOWRiNzM3ZTg3NTdhYTQ5OGQ0Mjg0ZWNhY2ZhNjA0MjJjYzdiYmM2MTdjZmFjNWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.PVKZEM7mSNFRPMxrwbwoe2pJuuACRUoLQfWHvo1Okbg)

Wow that was a shockingly good catch! I've removed all the changes to the Switch now. I was originally using it here, but ended up changing the design anyway so it's back to how it was before.

nabramow committed:
Adjust discover search filters

nabramow commented:
> One more quirk I noticed - it's unexpected to me that when you got a location filter on (by typing something in the input), then adding and removing "my communities" and "online" filters would reset it. See recording:
> 

@darrenvong I thiiiink I fixed this now. I believe the problem was I was clearing the search result after selecting the online only filter, but that wasn't clearing the visible search input value, so when you remove the filter it's gone but the text is still there.

For the life of me cannot find an accessible `onClear` function for the `LocationAutocomplete` component. Can't even see how that little clear `x` is getting there. So I just took out that invisible logic so at least what you see matches now.

Thank you for such a thorough test!

bakeiro commented:
> > I'm just wondering...
> > how the event location search works? its by area? (2 coordinates and then search ~50km around it?) if that's the case what happens when user types a country/state/region, would be only the center of that region? (circle area instead of a polygon)
> > if that's the case we should allow only filter by city... (or smaller 'nodes') I can create another ticket for this if it's needed
> > let me research about this and create a proper ticket... I already faced similar issues while working on the map
> 
> So if you look at `EventSearchReq` or you look at the backend function `def EventSearch` there's a few options for how search can work.
> 
> @bakeiro You can "search in rectangle", that's what I chose here. With than you give it a `RectArea` which is a latitude min, latitude max, longitude min and longitude max to search in. So you can calculate that by taking the exact location of the search coordinates and adding some "buffer" around them.
> 
> You can also "search in area", with that one you give a latitude, longitude and then a radius.
> 
> I don't think it's bad to also allow filter by country or state? What if you live in a tiny country like Monaco or something, haha, or a small US state. Maybe you're willing to drive a bit if you're not in a city? What's the downside or making it flexible?

The point is, I search "thailand" probably will look within an square, [so not respecting the frontiers], this can lead to overshow/undershow all the events wihtin that territory. We can solve this by using 'geometries' or polygons (or whatever are called) these polygons are available in open-street map, I was thinking about implementing then in map search (something that finally I didn't)

Without polygons, user will think "Oh I can search in a whole country/state" but in reality it's only either to small or to big

bakeiro commented:
![1](https://github.com/user-attachments/assets/8888b140-4cf5-4b22-9cee-cea72df7a78a)

this is what I meant, my suggestion here is:
In this iteration we should allow only the small 'nodes' (city, neighborhood, province), once we implement polygons search, we 
could search within any node regardless of its size.

Expected result when searching in a big node
![image](https://github.com/user-attachments/assets/b3790fdf-50e4-41a5-b76e-8ac87c6feb97)

I know how to implement the polygons, I can give you all the information if you need it.

bakeiro commented:
@nabramow 

![image](https://github.com/user-attachments/assets/ffdab23e-382d-4097-b964-58dfd55f20a9)
France even shows results within the US and south Spain.. but Portugal seems works good, and Spain works "OK"... I wonder how that rectagle is calculated 🤔😅

I mean, it's better than don't having it? idk I can create another ticket or you can limit search nodes in this PR as you wish :)

nabramow committed:
Fix region issue

nabramow commented:
>France even shows results within the US and south Spain.. but Portugal seems works good, and Spain works "OK"... I > > wonder how that rectagle is calculated 🤔😅
> 
> I mean, it's better than don't having it? idk I can create another ticket or you can limit search nodes in this PR as you wish :)

Oooo this was a great find, I didn't have the right test data to see it!

I've fixed this now. What I did: I checked if `isRegion = true`. If so, I search by query using location name (i.e. "France"). If it's not a region, I use the bbox rectangle search.

I tried it for Canada, Germany, France, Portugal, Thailand, United States. All seem to be working with my test data:

<img width="1028" alt="Screenshot 2024-10-23 at 8 13 47 PM" src="https://github.com/user-attachments/assets/dc47e451-0dd5-4e38-98a1-9c49040ddf89">

<img width="1028" alt="Screenshot 2024-10-23 at 8 14 00 PM" src="https://github.com/user-attachments/assets/548976f7-7dc6-4635-b7bb-16910dcf46d2">

<img width="1012" alt="Screenshot 2024-10-23 at 8 14 10 PM" src="https://github.com/user-attachments/assets/789e20e2-16a8-4f9d-bfc0-9bc531b1603d">

@bakeiro let me know if it works for you too!

bakeiro commented on app/web/service/search.ts:
```diff
@@ -98,12 +98,20 @@ export async function EventSearch({
+    // If it's a region (i.e. "France" or "United States") use query search by name
+    // This will search for events in the region by that name
+    if (searchLocation?.isRegion) {
+      req.setQuery(new StringValue().setValue(searchLocation.name));
```
this works by I'm trying to understand why... this simply does something like a regex? like if the searched name is inside the event location name?

bakeiro replied:
explain before merging pls, might be useful for the map in the future (even if not want to understand it :D)

nabramow replied:
Sure! It goes like this:

- We type a region like "France" in the search box
- That gets set in `LocationAutocomplete` component
- `LocationAutocomplete` component sends it in an api call to openstreetmap via the `useGeocodeQuery` hook
- Openstreetmap processes it, and gives us back the `GeocodeResult` object, standardizing the name, simplified name, bbox, lat and longitude, etc.
- We send that to our backend to to the `EventSearch` function to try to find matching events
- The backend checks what type of search: `search_in_area`, `search_in_rectangle` or `query` in this case. We define which search in the `search.ts` file with a grpc request
- We decide what to pass in the query. I am passing the location `name` passed back from the openstreet map response. We could also send `simplifiedName`, or something else.
- The backend checks if we want to search only the event title, or the event title, content AND address
- I chose the one with title, content and address
- Our event addresses have the city, state, country in them, that's why it works to search by country or city, as that's always in there. When you create an event, it also standardizes it with `LocationAutocomplete` via the openstreetmap api, so it's pretty safe that it will always match. So no chance someone makes a typo, leaves out the city or country, etc.

<img width="381" alt="Screenshot 2024-10-24 at 7 30 02 PM" src="https://github.com/user-attachments/assets/0a02365a-839f-4843-b2d4-1c868d6b283a">

The ONLY thing I could think that might go wrong...since the query also searches the content of the event, if the user adds the name of a different city in the event description, that might make it show up in a search.

So say I make an event in Paris, France, but in the description I write, "David and I met in Madrid and decided to organize this event together", the event might also show up when searching Madrid.

I think that's a rare edge case though, so it's probably okay.

Looks like @spreeni wrote this code on the backend, please correct me if I got anything wrong!

bakeiro commented:
ok, with Valencia from Spain works and even from Valencia from Venezuela works... [they don't get mixed] I still don't understand how that `setQuery` works but it works
